File: internal/proxy/fallback.go
package proxy

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"sync"

	"github.com/Open-Copilot-Proxy/Copilot_Proxy/internal/config"
)

type FallbackSelector struct {
	mu         sync.RWMutex
	cfg        config.Config
	httpClient *http.Client
	selected   string
}

func NewFallbackSelector(cfg config.Config, client *http.Client) *FallbackSelector {
	if client == nil {
		client = http.DefaultClient
	}
	return &FallbackSelector{cfg: cfg, httpClient: client}
}

func (s *FallbackSelector) Selected() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.selected
}

func (s *FallbackSelector) Set(model string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.selected = model
}

func (s *FallbackSelector) UpdateConfig(cfg config.Config) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.cfg = cfg
	s.selected = ""
}

func (s *FallbackSelector) Choose(ctx context.Context, modelsURL string, headers map[string]string) (string, error) {
	s.mu.RLock()
	if s.selected != "" {
		selected := s.selected
		s.mu.RUnlock()
		return selected, nil
	}
	cfg := s.cfg
	httpClient := s.httpClient
	s.mu.RUnlock()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, modelsURL, nil)
	if err != nil {
		return "", err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("models endpoint returned %d", resp.StatusCode)
	}
	var payload any
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return "", err
	}
	items := extractModelItems(payload)
	usableItems := make([]map[string]any, 0, len(items))
	for _, item := range items {
		if isEnabled(item) && isPickerEnabled(item) && supportsEndpoint(item, cfg.Fallback.RequiredEndpoint) {
			usableItems = append(usableItems, item)
		}
	}
	for _, pref := range cfg.Fallback.PreferredPrefixes {
		for _, item := range usableItems {
			if id, _ := item["id"].(string); id != "" && id == pref {
				s.Set(id)
				return id, nil
			}
		}
		for _, item := range usableItems {
			if matchesPrefix(item, pref) {
				id, _ := item["id"].(string)
				if id == "" {
					continue
				}
				s.Set(id)
				return id, nil
			}
		}
	}
	for _, item := range usableItems {
		id, _ := item["id"].(string)
		if id == "" {
			continue
		}
		s.Set(id)
		return id, nil
	}
	return "", nil
}

func extractModelItems(payload any) []map[string]any {
	if root, ok := payload.(map[string]any); ok {
		if data, ok := root["data"].([]any); ok {
			return mapsFromAny(data)
		}
		if models, ok := root["models"].([]any); ok {
			return mapsFromAny(models)
		}
	}
	if items, ok := payload.([]any); ok {
		return mapsFromAny(items)
	}
	return nil
}

func mapsFromAny(items []any) []map[string]any {
	result := make([]map[string]any, 0, len(items))
	for _, item := range items {
		if m, ok := item.(map[string]any); ok {
			result = append(result, m)
		}
	}
	return result
}

func isEnabled(model map[string]any) bool {
	if id, _ := model["id"].(string); strings.HasPrefix(id, "claude-opus-") {
		return false
	}
	policy, _ := model["policy"].(map[string]any)
	state, _ := policy["state"].(string)
	// GitHub API 返回多种状态：enabled, disabled, beta, hidden, expired
	// 只有 disabled 明确不可用，其他状态都视为可用
	if state == "" || state == "enabled" || state == "beta" || state == "hidden" {
		return true
	}
	return false
}

func isPickerEnabled(model map[string]any) bool {
	enabled, ok := model["model_picker_enabled"].(bool)
	return !ok || enabled
}

func supportsEndpoint(model map[string]any, required string) bool {
	if required == "" {
		return true
	}
	raw, ok := model["supported_endpoints"].([]any)
	if !ok || len(raw) == 0 {
		return true
	}
	for _, item := range raw {
		if endpoint, ok := item.(string); ok && endpoint == required {
			return true
		}
	}
	return false
}

func matchesPrefix(model map[string]any, pref string) bool {
	pref = strings.ToLower(pref)
	for _, key := range []string{"id", "version", "name", "family"} {
		if value, ok := model[key].(string); ok && strings.HasPrefix(strings.ToLower(value), pref) {
			return true
		}
	}
	return false
}

File: fallback.py
from typing import Any

import requests

PREFERRED_PREFIXES: list[str] = [
    'gpt-4.1',
    'gpt-4o',
    'gpt-5-mini',
    'raptor-mini',
]

def _is_enabled(model: dict[str, Any]) -> bool:
    policy = model.get('policy', {}) or {}
    state = policy.get('state')
    return state != 'disabled'

def _is_picker_enabled(model: dict[str, Any]) -> bool:
    return model.get('model_picker_enabled', True) is not False

def _supports_endpoint(model: dict[str, Any], required_endpoint: str | None) -> bool:
    endpoints = model.get('supported_endpoints')
    if not required_endpoint:
        return True
    if not endpoints:
        return True
    return required_endpoint in endpoints

def _extract_items(payload: Any) -> list[dict[str, Any]]:
    # 兼容不同模型列表返回结构。
    if isinstance(payload, dict):
        data = payload.get('data')
        if isinstance(data, list):
            return data
        models = payload.get('models')
        if isinstance(models, list):
            return models
    if isinstance(payload, list):
        return payload
    return []

def choose_fallback_model(
    models_url: str = 'http://127.0.0.1:15432/v1/models',
    headers: dict[str, str] | None = None,
    required_endpoint: str | None = '/chat/completions',
) -> str | None:
    """
    从本地代理的 /v1/models 中选择优先的 fallback 模型

    选择策略：
    1. 按 PREFERRED_PREFIXES 中的顺序，先尝试精确 id 匹配（例如 'gpt-4.1'），若未命中再按前缀匹配 id/version/name/family
    2. 如果未找到，则返回第一个未被 disabled 的 model['id']
    3. 找不到时返回 None
    """

    try:
        r = requests.get(models_url, headers=headers, timeout=5)
    except Exception as e:
        print(f'[D] choose_fallback_model: 请求失败 {e}')
        return None

    if r.status_code != 200:
        preview = (r.text or '')[:120].replace('\n', ' ')
        print(f'[D] choose_fallback_model: HTTP {r.status_code}, body={preview}')
        return None

    try:
        data = r.json()
    except Exception as e:
        preview = (r.text or '')[:120].replace('\n', ' ')
        print(f'[D] choose_fallback_model: JSON 解析失败 {e}, body={preview}')
        return None

    items = _extract_items(data)
    print(f'[D] choose_fallback_model: 获取到 {len(items)} 个模型')

    def matches_prefix(m: dict[str, Any], pref: str) -> bool:
        lower = pref.lower()
        for key in ('id', 'version', 'name', 'family'):
            v = m.get(key)
            if not v:
                continue
            if isinstance(v, str) and v.lower().startswith(lower):
                return True
        return False

    def is_usable(m: dict[str, Any]) -> bool:
        return _is_enabled(m) and _is_picker_enabled(m) and _supports_endpoint(m, required_endpoint)

    for pref in PREFERRED_PREFIXES:
        for item in items:
            if item.get('id') == pref and is_usable(item):
                print(f'[D] choose_fallback_model: 命中优先 id {pref}, 选择 {item.get("id")}')
                return item.get('id')

        for item in items:
            if matches_prefix(item, pref) and is_usable(item):
                print(f'[D] choose_fallback_model: 命中优先级前缀 {pref}, 选择 {item.get("id")}')
                return item.get('id')

    for item in items:
        if is_usable(item):
            print(f'[D] choose_fallback_model: 无优先级匹配，返回首个 enabled: {item.get("id")}')
            return item.get('id')

    print('[D] choose_fallback_model: 没有找到任何可用模型')
    return None

if __name__ == '__main__':
    print('Chosen fallback:', choose_fallback_model())
